decoders/ps2/pd.py

```diff
@@ -62,65 +83,89 @@ def reset(self):
 
     def start(self):
         self.out_ann = self.register(srd.OUTPUT_ANN)
-
-    def putb(self, bit, ann_idx):
-        b = self.bits[bit]
-        self.put(b.ss, b.es, self.out_ann, [ann_idx, [str(b.val)]])
-
-    def putx(self, bit, ann):
+        self.out_py = self.register(srd.OUTPUT_PYTHON)
+
+    def metadata(self,key,value):
+        if key == srd.SRD_CONF_SAMPLERATE:
+            self.samplerate = value
+
+    def get_bits(self, n, edge:'r', timeout=100e-6):
+        max_period = int(timeout * self.samplerate) + 1
+        for i in range(n):
+            _, dat = self.wait([{0:edge},{'skip':max_period}])
+            if not self.matched[0]:
+                break #Timed out
+            self.bits.append(Bit(dat, self.samplenum, self.samplenum+max_period))
+            if i>0: #Fix the ending period
+                self.bits[i-1].es = self.samplenum
+        if len(self.bits) == n:
+            self.wait([{0:'r'},{'skip':max_period}])
+            self.bits[-1].es = self.samplenum
+        self.bitcount = len(self.bits)
+
+    def putx(self, bit, ann, host=False):
+        if host:
+            ann[0] += 7 #host annotation offset
         self.put(self.bits[bit].ss, self.bits[bit].es, self.out_ann, ann)
 
-    def handle_bits(self, datapin):
-        # Ignore non start condition bits (useful during keyboard init).
-        if self.bitcount == 0 and datapin == 1:
-            return
-
-        # Store individual bits and their start/end samplenumbers.
-        self.bits.append(Bit(datapin, self.samplenum, self.samplenum))
-
-        # Fix up end sample numbers of the bits.
-        if self.bitcount > 0:
-            b = self.bits[self.bitcount - 1]
-            self.bits[self.bitcount - 1] = Bit(b.val, b.ss, self.samplenum)
-        if self.bitcount == 11:
-            self.bitwidth = self.bits[1].es - self.bits[2].es
-            b = self.bits[-1]
-            self.bits[-1] = Bit(b.val, b.ss, b.es + self.bitwidth)
-
-        # Find all 11 bits. Start + 8 data + odd parity + stop.
-        if self.bitcount < 11:
-            self.bitcount += 1
-            return
+    def handle_bits(self, host=False):
+        # Annotate individual bits
+        for b in self.bits:
+            self.put(b.ss, b.es, self.out_ann, [Ann.BIT, [str(b.val)]])
 
         # Extract data word.
         word = 0
-        for i in range(8):
+        for i in range(min(8, self.bitcount-1)):
```

Comment:
is this the correct ? I would've assumed bitcount would "usually" be 8, resolving into `range(7) `

Reply:
I was trying to prevent invalid index errors in the event that the host interrupts the client in the middle of a byte. The alternative is to guard with `if self.bitcount>8:` and not parse fewer than 8 bits.

Reply:
fair - I'm not familiar enough with this proto to double-check your calculations but I saw a lot of adjustments and checks to bit counters, and to me looks like a lot of potential for off-by-one errors - even more reason for a good test suite !